### PR TITLE
Add reactive NSPopUpButton

### PR DESCRIPTION
**Pull request by ohbargain**
- [x]  Add selected index signal
- [x]  Add selected index binding target
- [x]  Add selected title signal
- [x]  Add selected title binding target

### CODE REVIEW

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	public var selectedIndex: BindingTarget<Int> {
+		return makeBindingTarget { $0.selectItem(at: $1) }
+	}
+	
+	/// A signal of selected title
+	public var selectedTitles: Signal<String, NoError> {
+		return self.integerValues.map { [unowned base = self.base] _ -> String? in
+			return base.selectedItem?.title
+		}
+		.skipNil()
```
Comment (mdiep) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Can you format these in the RAC style?

```swift
return self.integerValues
	.map { [unowned base = self.base] _ -> String? in
		return base.selectedItem?.title
	}
	.skipNil()
```

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	
+	/// A signal of selected indexes
+	public var selectedIndexes: Signal<Int, NoError> {
+		return self.integerValues.map { [unowned base = self.base] _ -> Int? in
+			guard let item = base.selectedItem else { return nil }
+			return base.index(of: item)
+		}
+		.skipNil()
+	}
+	/// Sets the button with an index.
```
Comment (mdiep) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Missing a newline here.

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+			.map { [unowned base = self.base] _ -> String? in
+				return base.selectedItem?.title
+			}
+			.skipNil()
+	}
+	
+	/// Sets the button with title.
+	public var selectedTitle: BindingTarget<String> {
+		return makeBindingTarget { $0.selectItem(withTitle: $1) }
+	}
```
Comment (mdiep) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Looking at these again, I think they should probably all be optional—with `nil` meaning no item is selected. Sorry I didn't catch this before.

`NSPopUpButton` isn't entirely annotated correctly, but this is still doable. (Tests should be updated as well.)

Comment (ohbargain) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Yes, I agree. Will make the change.

> NSPopUpButton isn't entirely annotated correctly...

Could you be more specific? Are you referring to API availability annotations?

Comment (mdiep) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Yup. But an empty string also deselects everything. So we can convert nil to the empty string.

Comment (ohbargain) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Having attempted at this, NSPopUpButton is always selected and does not have option to allow for "not selected" state. 
This seems to be why `NSPopUpButton.indexOfSelectedItem` is not optional. 
Although `NSPopUpButton.titleOfSelectedItem` is optional, the underlying property `NSMenuItem.title` is not optional.
All of this makes setting and emitting nil values impossible, thus untestable. I don't think these APIs should be optional.

**Comment by ohbargain** [comment]
Programmatically, it is possible to make `NSPopUpButton` in nil state. So making BindingTarget optional make sense and I have made the changes accordingly.  But I can't figure out how to invoke this as user action, making it impossible to test.  If there are any ways, I would love to know.  
Otherwise, perhaps we should make emitting signal non-optional?

**Comment by mdiep** [comment]
> Otherwise, perhaps we should make emitting signal non-optional?

Yeah, that makes sense. 👍

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import ReactiveSwift
+import AppKit
```
Comment (tbaranes) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Inieldizi
```
```

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+			.map { [unowned base = self.base] _ -> Int in
+				return base.indexOfSelectedItem
+			}
+	}
+	
+	/// Sets the button with an index.
+	public var selectedIndex: BindingTarget<Int?> {
+		return makeBindingTarget {
+			$0.selectItem(at: $1 ?? -1)
+		}
```
Comment (tbaranes) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: *
```oaks'
``` ´
```
```dziskksz*
``` 
```
```iziizzkso
```s
```´ i

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+extension Reactive where Base: NSPopUpButton {
+	
+	/// A signal of selected indexes
+	public var selectedIndexes: Signal<Int, NoError> {
+		return self.integerValues
+			.map { [unowned base = self.base] _ -> Int in
+				return base.indexOfSelectedItem
+			}
+	}
+
```
Comment (tbaranes) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: *
```oaks'
``` ´
```
```dziskksz*
``` 
```
```iziizzkso
```s
```´ i

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import ReactiveSwift
+import AppKit
+import enum Result.NoError
+
+extension Reactive where Base: NSPopUpButton {
+	
+	/// A signal of selected indexes
+	public var selectedIndexes: Signal<Int, NoError> {
+		return self.integerValues
```
Comment (tbaranes) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: *
```oaks'
``` ´
```
```dziskksz*
``` n
```w slack is_
```iziizzksozi zioeki
```s
```´ i

File: ReactiveCocoa/AppKit/NSPopUpButton.swift [hunk 8]
```diff
@@ -0,0 +1,38 @@
+import ReactiveSwift
+import AppKit
+import enum Result.NoError
+
+extension Reactive where Base: NSPopUpButton {
```
Comment (tbaranes) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Dla dsl' 
```nw'sks
```

Comment (tbaranes) [inline ReactiveCocoa/AppKit/NSPopUpButton.swift]: Dla dsl' 
```nw'sks
```

**Comment by andersio** [comment]
Mind to rebase on `master`?

**Comment by mdiep** [comment]
> Mind to rebase on master?

Why?

**Comment by andersio** [comment]
@mdiep Failing test.

**Comment by mdiep** [comment]
_facepalm_

**Comment by andersio** [comment]
OMG. 🤕

**Pull request merged**